File: libraries/OclProcess.cs

    interface Runnable
    { public void run(); }


    class OclProcess : Runnable
    {
        string name = "";
        int priority = 5;
        Runnable executes = null;
        Thread actualThread = null;
        Process osProcess = null;

        long deadline = 0;
        long delay = 0;
        long period = 0; 

        public static OclProcess newOclProcess(Runnable obj, string s)
        {
            OclProcess p = new OclProcess();
            p.executes = obj; 
            p.name = s;
            if (obj != null)
            {
                ThreadStart st = new ThreadStart(obj.run);
                p.actualThread = new Thread(st);
            }
            else
            {
                Process proc = new Process();
                // proc.StartInfo.UseShellExecute = false;
                proc.StartInfo.FileName = s;
                // proc.StartInfo.CreateNoWindow = true;
                p.osProcess = proc;
            }
            return p;
        }

        public void setDeadline(long d)
        { deadline = d; }

        public void setDelay(long d)
        { delay = d; }

        public void setPeriod(long p)
        { period = p; }

        public long getDeadline()
        { return deadline; }

        public static void notify(object obj)
        { Monitor.Pulse(obj); }

        public static void notifyAll(object obj)
        { Monitor.PulseAll(obj); }

        public static void wait(object obj, int n)
        {
            Monitor.Wait(obj, n);
            // var pause = Task.Delay(n);
            // pause.Wait(); 

            // int tot = (int) ((m * 1000000 + n) / 1000000.0);
            // var pause = Task.Delay(tot);
            // pause.Wait();
        }

        public int waitFor()
        {
            if (osProcess != null)
            {
                osProcess.WaitForExit();
                return 0;
            }
            return 0;
        }

        public static OclProcess currentThread()
        {
            Thread t = Thread.CurrentThread;
            OclProcess pr = new OclProcess();
            pr.name = "Current thread";
            pr.actualThread = t;
            return pr;
        }


        public static OclProcess getRuntime()
        { return OclProcess.currentThread(); }

        public static ArrayList allActiveThreads()
        { return new ArrayList(); }

        public static int activeCount()
        { return ThreadPool.ThreadCount; }

        public static void sleep(long n)
        { Thread.Sleep((int)n); }

        public string getName()
        { return name; }

        public void setName(string nme)
        { name = nme; }

        public int getPriority()
        {
            if (actualThread != null)
            { return (int)actualThread.Priority; }
            return priority;
        }

        public bool isAlive()
        {
            if (actualThread == null)
            { return false; }
            return actualThread.IsAlive;
        }

        public bool isDaemon()
        {
            if (actualThread == null)
            { return false; }
            return actualThread.IsBackground;
        }

        public void run()
        { if (executes is OclProcess)
          {
            OclProcess proc = (OclProcess)executes;
            if (proc.delay > 0)
            { Thread.Sleep((int)proc.delay); }
                if (proc.period > 0)
                {
                    long now = SystemTypes.getTime();
                    long next = now + proc.period;
                    while (true)
                    { proc.run();
                      now = SystemTypes.getTime();
                      next = next + proc.period;
                      if (next > now)
                      { Thread.Sleep((int)(next - now)); }
                      ThreadStart st = new ThreadStart(proc.executes.run);
                      proc.actualThread = new Thread(st);
                    }
                }
                else
                { proc.run(); }

            return; 
          }

            if (actualThread != null)
            { actualThread.Start(); }
            if (osProcess != null)
            { osProcess.Start(); }
        }

        public void start()
        {
            if (actualThread != null)
            { actualThread.Start(); }
            if (osProcess != null)
            { osProcess.Start(); }
        }

        public void join(double ms)
        {
            if (actualThread != null)
            { actualThread.Join((int)ms); }
        }

        public void interrupt()
        {
            if (actualThread != null)
            { actualThread.Interrupt(); }
        }

        public void cancel()
        { destroy(); }

        public void destroy()
        {
            if (actualThread != null)
            {
                actualThread.Abort();
                actualThread = null;
            }

            if (osProcess != null)
            {
                osProcess.Kill();
                osProcess = null;
            }
        }

        public static string getEnvironmentProperty(string var)
        { return System.Environment.GetEnvironmentVariable(var); }

        public static Hashtable getEnvironmentProperties()
        { IDictionary dict = System.Environment.GetEnvironmentVariables(); 
          Hashtable res = new Hashtable(); 
          foreach (object x in dict.Keys)
          { res[x] = dict[x]; }
          return res;
        }

        public static string setEnvironmentProperty(string var, string val)
        { string prop = 
              System.Environment.GetEnvironmentVariable(var);  
          System.Environment.SetEnvironmentVariable(var,val); 
          return prop; 
        }

        public static string clearEnvironmentProperty(string var)
        { string prop = 
              System.Environment.GetEnvironmentVariable(var);  
          System.Environment.SetEnvironmentVariable(var,null); 
          return prop; 
        }

        public static void exit(int n)
        { System.Environment.Exit(n); }
    }

File: libraries/oclprocess.py
import threading
import time
import os


class OclProcess:
  oclprocess_instances = []
  oclprocess_index = dict({})

  def __init__(self):
    self.name = ""
    self.priority = 0
    self.actualThread = None
    self.executes = None
    self.osProcess = 0
    self.deadline = 0
    self.delay = 0
    self.period = 0
    OclProcess.oclprocess_instances.append(self)


  def notify(obj) :
    obj.notify()

  def notifyAll(obj) :
    obj.notify_all()

  def wait(obj,t) :
    obj.wait(t/1000.0)

  def waitFor(self) :
    if self.osProcess > 0 : 
      (x,y) = os.waitpid(self.osProcess,0)
      return y
    return 0

  def getRuntime() : 
    return OclProcess.currentThread()

  def newOclProcess(obj,s) :
    result = None
    p = createOclProcess()
    p.name = s
    if obj != None : 
      p.actualThread = threading.Thread(target=obj,name=s)
      p.executes = obj
    else : 
      p.actualThread = None
      pid = os.spawnv(os.P_NOWAIT, s, ["."])
      p.osProcess = pid
    p.priority = 5
    result = p
    return result

  def getDeadline(self) : 
    return self.deadline

  def setDeadline(self,d) : 
    self.deadline = d

  def setDelay(self,d) : 
    self.delay = d

  def setPeriod(self,p) :
    self.period = p

  def setPriority(self,p) : 
    self.priority = p

  def activeCount() :
    return threading.active_count()

  def currentThread() :
    return threading.current_thread()

  def allActiveThreads() : 
    return threading.enumerate()

  def sleep(n) :
    time.sleep(n/1000.0)

  def sleepSeconds(d) :
    time.sleep(d)

  def getName(self) : 
    result = self.name
    if self.actualThread != None : 
      result = self.actualThread.name
    return result

  def setName(self,nme) : 
    self.name = nme
    if self.actualThread != None : 
      self.actualThread.name = nme
    return result

  def getPriority(self) :
    return self.priority

  def run(self) :
    if self.actualThread != None : 
      if hasattr(self.executes, "deadline") and hasattr(self.executes, "delay") and hasattr(self.executes, "period") and self.executes.deadline == 0 and self.executes.delay == 0 and self.executes.period == 0 :
        self.actualThread.run()
      elif hasattr(self.executes, "deadline") and hasattr(self.executes, "delay") and hasattr(self.executes, "period"): 
        now = time.time()*1000
        if self.executes.deadline > now :
          time.sleep((self.executes.deadline - now)/1000)  
        if self.executes.delay > 0 :
          time.sleep(self.executes.delay/1000) 
        if self.executes.period <= 0 : 
          self.actualThread.run()
        else : 
          next = now + self.executes.period
          while True : 
            self.actualThread.run() 
            now = time.time()*1000
            next = next + self.executes.period  
            if next > now :
              time.sleep((next - now)/1000)
            self.actualThread = threading.Thread(target=self.executes,name=self.name)
      else : 
        self.actualThread.run()

  def start(self) :
    if self.actualThread != None : 
      if hasattr(self.executes, "deadline") and hasattr(self.executes, "delay") and self.executes.deadline == 0 and self.executes.delay == 0 :
        self.actualThread.start()
      elif hasattr(self.executes, "deadline") and hasattr(self.executes, "delay") : 
        now = time.time()*1000
        if self.executes.deadline > now :
          time.sleep((self.executes.deadline - now)/1000)  
        if self.executes.delay > 0 :
          time.sleep(self.executes.delay/1000) 
        self.actualThread.start()
      else : 
        self.actualThread.start()

  def startProcess(self) : 
    self.start()
    return self

  def isAlive(self) : 
    if self.actualThread != None : 
      return self.actualThread.is_alive()
    return False

  def isDaemon(self) : 
    if self.actualThread != None : 
      return self.actualThread.daemon
    return False

  def join(self,ms) :
    if self.actualThread != None : 
      self.actualThread.join(ms/1000.0)

  def destroy(self) :
    if self.actualThread != None : 
      del self.actualThread
      self.actualThread = None

  def interrupt(self) :
    pass

  def cancel(self) :
    pass

  def getEnvironmentProperty(var) : 
    if var in os.environ : 
      return os.environ[var]
    return ""

  def getEnvironmentProperties() : 
    props = os.environ
    res = dict([])
    for k in props : 
      res[k] = props[k]
    return res

  def setEnvironmentProperty(var,value) : 
    res = ""
    if var in os.environ : 
      res = os.environ[var]
    os.environ[var] = value
    return res

  def clearEnvironmentProperty(var) : 
    res = ""
    if var in os.environ : 
      res = os.environ[var]
    del os.environ[var]
    return res

  def exit(n) : 
    exit(n)

  def killOclProcess(oclprocess_x) :
    oclprocess_instances = ocl.excludingSet(oclprocess_instances, oclprocess_x)
    free(oclprocess_x)

def createOclProcess():
  oclprocess = OclProcess()
  return oclprocess

def allInstances_OclProcess():
  return OclProcess.oclprocess_instances


# p = OclProcess.newOclProcess(None, "C:/Program Files/Windows NT/Accessories/wordpad.exe")
# print("Process id = " + str(p.osProcess))
# ec = p.waitFor()
# print(ec)

# print(OclProcess.getEnvironmentProperties())
# OclProcess.setEnvironmentProperty("alpha", "C:\\alpha")
# print(OclProcess.getEnvironmentProperty("alpha"))
